### ISSUE TITLE
How to extend action message ?

### ISSUE THREAD

**Issue by wwj718** [operand/agency#95]
I want to add a field for action: message_id,  used to match sent and received messages. Other potential extension requirements include timestamp...

Currently I seem to need to modify the internal implementation of Agent (MessageSchema or ActionSchema). Is there a better way ?

**Comment by operand** [comment]
You'd have to make a change to the implementation unfortunately. I also want to add a message id field.

If you saw the other issue regarding action parameter descriptions, I proposed adding an optional field called `__meta__` that we can use to store this kind of metadata.

Copying from that issue, here's an example of what I'm thinking with a message id and timestamp in the meta field. (note that I moved the meta field out of the action field in this example. I think this might be a better format.)

```json
{
  "to": "John",
  "from": "Jane",
  "__meta__": {
    "message_id": "abc123",
    "timestamp": "2023-08-01 01:00 UTC",
    "thoughts": "I should greet John",
  },
  "action": {
    "say": {
      "content": "Hello, John!"
    }
  }
}
```

I'm thinking we could change `Space._route` to add a message id to the meta field automatically. Here's how I'm thinking you would then send a message:

```python
sent_message = self._send({
  "to": "John",
  "__meta__": {
    "thoughts": "I should greet John",
  },
  "action": {
    "say": {
      "content": "Hello, John!"
    }
  }
})
```

Not much of a difference except that it would return the full message as it was sent, including any additional meta fields and the `from` field populated. So `sent_message` would then contain:

```python
sent_message == {
  "to": "John",
  "from": "Jane", # this is added by the _route method
  "__meta__": {
    "message_id": "abc123", # this is added by the _route method
    "timestamp": "2023-08-01 01:00 UTC", # this is added by the _route method
    "thoughts": "I should greet John",
  },
  "action": {
    "say": {
      "content": "Hello, John!"
    }
  }
}
```

Note that you don't need to store `sent_message` since it will automatically be added to the `_message_log` list, but this would allow you to inspect the complete message if needed after sending. So you could pull out the `message_id` field or others if you need to.

Then we can also change how the `return` and `error` actions work, to use the message id instead of sending the entire original message (which is a flawed approach). Here's how a `return` message would look:

```python
{
  "to": "Jane",
  "from": "John",
  "__meta__": {
    "message_id": "def456", # this is the id for the `return` message
  },
  "action": {
    "return": {
      "original_message_id": "abc123", # Note the original message id here
      "return_value": "Hi, Jane!",
    }
  }
}
```

Then I think the `return` and `error` actions wouldn't need to change. You'd still use the same signature:

```python
def _action__return(self, original_message: dict, return_value: str):
    ...
```

But the `original_message` field would be determined by the library by automatically correlating message id's. And we could add a method for doing this manually when needed. Something like:

```python
def _find_message(self, message_id: str) -> dict:
    """Returns the message with the given id."""
```

What do you think of this? I'm in the middle of implementing a lot of these changes now. I hope to have a 1.3 out in a few days and I hope to include something like this.

**Comment by wwj718** [comment]
I like the idea, "__meta__" is very useful for extensions.

I have some opinions about the fields of the route automatic operation:

```js
sent_message == {
  "to": "John",
  "from": "Jane", # this is added by the _route method
  "__meta__": {
    "message_id": "abc123", # this is added by the _route method
    "timestamp": "2023-08-01 01:00 UTC", # this is added by the _route method
    "thoughts": "I should greet John",
  },
  "action": {
    "say": {
      "content": "Hello, John!"
    }
  }
}
```

- from: I have a use case: there is an Agent  work like a message transfer station, which will forward message from another system (non-Python system, Agent cannot be implemented internally), when it forwards the message, it does not want to modify the from field.  
- message_id:  The Actor model is inherently an asynchronous system, sometimes we need to implement synchronous semantics. Interestingly, we can implement it only on the client side of the message sent(This is a proof of Actor's flexibility, synchronization semantics can be built on it, and it is only implemented inside a certain Actor, the system remains ignorant of it)，the following is an example
        
```python
class SyncAgent(Agent):
    '''
    a synchronous Agent
    '''

    def __init__(self, id):
        super().__init__(id)
        self._future_resolves = {}

    def _receive(self, message):
        threading.Thread(target=super()._receive, args=(message,)).start()
        
    @access_policy(ACCESS_PERMITTED)
    def _action__return(self, original_message: dict, return_value):
        if original_message.get("message_id"):
            in_comming_message_id =  original_message['message_id']
            future = self._future_resolves.get(in_comming_message_id)
            if future:
                future.set_result(return_value)

    def _send_and_wait(self, message, timeout=None):
        with ThreadPoolExecutor(max_workers=1) as executor:
            future = Future()
            task = executor.submit(lambda f: f.result(timeout=timeout), future)
            message_id = random.randint(1, 100000000)
            self._future_resolves[message_id] = future # todo: threadsafe
            message["message_id"] = message_id
            self._send(message)
            return task.result() # block and wait the result or TimeoutError

    @access_policy(ACCESS_PERMITTED)
    def _action__test(self, content: str):
        """test"""
        message = {
            "to": "John",
            "thoughts": "",
            "action": "test",
            "args": {
                "content": f"hi {random.randint(1,10000)}"
            },
        }

        result = self._send_and_wait(message)
        print("result: ", result)
```

This SyncAgent generates a message_id on the Agent side, immediately enters the block after sending a message, and waits (or times out) for a message with the same message_id, which requires getting the message_id immediately after sending a message. If message_id is sent by _route, can `_send` method return message_id as return value

Based on the above two use cases, it seems that the agent can directly manipulate __meta__ to bring flexibility, which raises the question: which behavior of _route should be mandatory, or should the message from the agent be transmitted as transparently as possible.

**Comment by wwj718** [comment]
It is also worth noting here:

```python
    def _receive(self, message):
        threading.Thread(target=super()._receive, args=(message,)).start()
```

The current execution mechanism of Agent action is blocking. It needs to wait for an action to complete before receiving and executing the next message.  We can decide the concurrency mechanism of action execution inside the Agent(like SyncAgent). But it's not clear to me that there is value in adding this mechanism as an option to the base class.

**Comment by operand** [comment]
I'm really glad that you're bringing these things up. I think we're very aligned in how we're thinking about all this. I've wondered about many of these choices and it's really helpful to hear your perspective.

On the added fields: `from`, `message_id`, etc. I think you bring up a really good point. I like the idea of moving this responsibility entirely to the Agent class like you're suggesting.

This way the Space classes would be entirely focused on routing, based on the `to` and `from` fields and it would not affect anything in the message.


I also love the synchronous messaging example you gave. I've had that thought as well and I think it supports the idea of having the agent determine all message fields. This way the entire implementation of synchronous messaging could be in the agent class. This seems like it would allow more flexibility overall.

So I'm going to include this in the coming changes. All the fields will need to be set by an Agent. `Space._route` would only be interested in the `to` and `from` fields for routing but would not touch anything in the message.


And regarding the `_receive` method and whether it should parallelize processing automatically. I like that possibility a lot but I'm not confident I want to add that yet. Mainly because I've got a big set of updates already and I'd like to think about this some more. But I agree this could be a very useful option. I'm going to add this as a note under the [Improve Multiprocessing](https://github.com/operand/agency/issues/33) issue so that it isn't forgotten. Maybe after the next release we can play around on a PR for this.

Let me know if this all sounds good to you!

**Comment by wwj718** [comment]
> So I'm going to include this in the coming changes. All the fields will need to be set by an Agent. Space._route would only be interested in the to and from fields for routing but would not touch anything in the message.

Awesome !  

I think this will better support [Late_binding](https://en.wikipedia.org/wiki/Late_binding), let the decision-making be postponed to the object (Agent) as much as possible. [Alan Kay](https://en.wikipedia.org/wiki/Alan_Kay) once said:

> OOP to me means only messaging, local retention, and protection and hiding of state-process, and extreme late-binding of all things.

The Actor model is very similar to [what Alan Kay did](http://stereobooster.github.io/two-big-schools-of-object-oriented-programming) in [Smalltalk](https://wikipedia.org/wiki/Smalltalk) in the early days([Joe Armstrong & Alan Kay](https://www.youtube.com/watch?v=fhOHn9TClXY)).   

It is an excellent decision for agency to choose Actor as the calculation model.

> I also love the synchronous messaging example you gave. 
> And regarding the _receive method and whether it should parallelize processing automatically. I like that possibility a lot but I'm not confident I want to add that yet. 

I am not sure about the concurrency and synchronization of Actors. I only have some rough ideas at present,  Erlang / [Elixir](https://elixir-lang.org/getting-started/processes.html#state) seems like a good place to learn, but I currently only have a rough understanding of them. 

I currently think about [the behaviors that the Agent should support](http://www.fipa.org/specs/fipa00037/SC00037J.html) in the context of [FIPA](https://en.wikipedia.org/wiki/Foundation_for_Intelligent_Physical_Agents). What are your plans for the behaviors that the Agent will support?

**Comment by operand** [comment]
Thank you! Interestingly, I didn't choose the actor model directly. I developed the basics of the framework iterating on my own agent system, and at one point I realized that I was approaching the Actor model. That's when the idea became clearer, and soon after I decided to open source.

Regarding FIPA, I'm not familiar but you gave me some good reading to do. :)

I'm aware of other similar "plug-and-play" like protocols that this resembles but I don't have a specific protocol I'm thinking of. I'm somewhat taking a "let's justify what we need as we go" approach and trying to minimize assumptions. That said I do like to review prior work to see how others have done it.

Overall I want to support practical development of agent systems. Since it's not entirely clear what the best approaches are, or that there is a "best" approach, I think experimentation is necessary at this time. So I want the library to be flexible but I also want to support developing real production systems once you've found an approach that works for you.

So I don't have a concrete set of features beyond what's on the readme or the issues page, but I look at those features as just a first milestone. Beyond that I have a number of ideas and things I'd like to include but I'm open to anything that would fit the goal of enabling others to develop their own agent systems.

**Comment by wwj718** [comment]
> I'm somewhat taking a "let's justify what we need as we go" approach and trying to minimize assumptions.

I like this idea of "let's justify what we need as we go" !

> Overall I want to support practical development of agent systems. Since it's not entirely clear what the best approaches are, or that there is a "best" approach, I think experimentation is necessary at this time. So I want the library to be flexible but I also want to support developing real production systems once you've found an approach that works for you.

Fully agree and support this way of exploring.

**Comment by operand** [comment]
This issue has been addressed with the 1.3 changes (https://github.com/operand/agency/pull/97). Specifically, you can now add a `meta` field to all messages. See: https://github.com/operand/agency/tree/main/docs#using-the-meta-field

Closing for now but feel free to reopen if needed.
